alphazero/mcts/mcts.py

```diff
@@ -307,7 +345,69 @@ def forward(root_node: MCTSNode):
     # TODO Maybe just iterate through children once, more efficient?
     # Currently doing it twice due to output difference,
     # just do forloop instead?
+    # This can def be optimized better
     return (
-        sum([child.action * child.n_visit for child in start_node.children]),
+        torch.pow(
+            sum(
+                [child.action * child.n_visit for child in start_node.children]
+            ),
+            (1 / temperature),
+        ),
         sampled_node,
     )
+
+
+def self_play(
+    game: GameClassType,
+    net: torch.nn.Module,
+    board_converter: BoardConverterType,
+    n_mcts_iter: int = 1600,
+    temperature: float = 1.0,
+    dirichlet_eps: float = 0.25,
+    dirichlet_conc: float = 1.0,
+) -> List[tuple]:
+    """Plays games with itself
+
+    Args:
+        game (GameClassType): The game class for the model
+        net (torch.nn.Module): The model network
+        board_converter (BoardConverterType, optional): Converts board state
+        to input tensor.
+        n_mcts_iter (int, optional): Number of mcts iterations.
+        Defaults to 1600.
+        temperature (float, optional): Temperature for probability calculation.
+        Defaults to 1.0.
+        dirichlet_eps (float, optional): Dirichlet noise epsilon.
+        Defaults to 0.25.
+        dirichlet_conc (float, optional): Dirichlet noise concentration.
```

Comment:
what is this exactly?

Reply:
Dirichlet noise that is added to the prior probability used for puct score in MCTS. It's mentioned in the AlphaZero paper